outrank/algorithms/synthetic_data_generators/cc_generator.py

```diff
@@ -0,0 +1,849 @@
+        :param y: required target labels for categorical noise generation
+        :param p: amount of noise to apply. Defaults to 0.2
+        :param type: type of noise to apply, either categorical or missing
+        :param missing_val: value to simulate missing values. Defaults to float('-inf')
+        :return: X with noise applied
+        """
+
+        self.dataset_info['noise'].append({
+            'type': type,
+            'amount': p,
+        })
+
+        if type == 'categorical':
+            label_values, label_count = np.unique(y, return_counts=True)
+            n_labels = len(label_values)
+
+            inds = y.argsort()
+            y_sort = y[inds]
+            X_sort = X[inds]
+
+            Xs_T = X_sort.T
+            n = Xs_T.shape[1]
+            n_flip = int(n * p)
+
+            for feature in Xs_T:
+                unique_per_label = {}
+
+                for i in range(n_labels):
+                    if i == 0:
+                        unique = np.unique(feature[:label_count[i]])
+                        unique_per_label[label_values[i]] = set(unique)
+                    else:
+                        unique = np.unique(feature[label_count[i - 1]:label_count[i - 1] + label_count[i] - 1])
+                        unique_per_label[label_values[i]] = set(unique)
+
+                ixs = np.random.choice(n, n_flip, replace=False)
+
+                for ix in ixs:
+                    current_label = y_sort[ix]
+                    possible_labels = np.where(label_values != current_label)[0]
+
+                    # find all unique values from labels != current label
+                    values = set()
+                    for key in possible_labels:
+                        values = values.union(unique_per_label[key])
+
+                    # remove any overlapping values, ensuring replacement values are unique & from a target label !=
+                    # current label
+                    for val in unique_per_label[current_label] & values:
+                        values.remove(val)
+
+                    if len(values) > 0:
+                        val = np.random.choice(list(values))
+
+                    else:
+                        key = possible_labels[np.random.randint(len(possible_labels))]
+                        values = unique_per_label[key]
+                        val = np.random.choice(list(values))
+
+                    feature[ix] = val
+
+            rev_ind = inds.argsort()
+            X_noise = Xs_T.T
+            X_noise = X_noise[rev_ind]
+
+            return X_noise
+
+        elif type == 'missing':
+            X_noise = np.copy(X)
+            Xn_T = X_noise.T
+            n = Xn_T.shape[1]
+            n_missing = int(n * p)
+            #print("n to delete:", n_missing)
+
+            for feature in Xn_T:
+                ixs = np.random.choice(n, n_missing, replace=False)
+
+                for ix in ixs:
+                    feature[ix] = missing_val
```

Comment:
You in general do a lot of for loop-based indexing -- this is really not the optimal way, this can all be vectorized. Not in this PR thought, let's go over separately sometime